cranelift/codegen/src/verifier/mod.rs

```diff
@@ -1124,6 +1148,66 @@ impl<'a> Verifier<'a> {
         }
     }
 
+    /// Checks If a `load` or `store` access is within `heap_addr` bounds.
+    fn verify_heap_addr_bounds(
+        &self,
+        inst: Inst,
+        opcode: Opcode,
+        addr: Value,
+        offset: Offset32,
+        errors: &mut VerifierErrors,
+    ) -> VerifierStepResult<()> {
+        let dfg = &self.func.dfg;
+
+        // First, let's check if `addr` comes from a `heap_addr` instruction.
+        let valid_size = match dfg.value_def(addr) {
+            ValueDef::Result(def_inst, _) => match dfg[def_inst] {
+                ir::InstructionData::HeapAddr { imm, .. } => u32::from(imm) as i64,
+                // Not a `heap_addr`, we can't verify this.
+                _ => return Ok(()),
+            },
+            _ => return Ok(()),
+        };
+
+        // Some of these instructions do a load + extend, which means
+        // they access less memory than the destination control type.
+        let access_size = match opcode {
+            Opcode::Istore8 | Opcode::Uload8 | Opcode::Sload8 => 1,
+            Opcode::Istore16 | Opcode::Uload16 | Opcode::Sload16 => 2,
+            Opcode::Istore32 | Opcode::Uload32 | Opcode::Sload32 => 4,
+            Opcode::Uload8x8
+            | Opcode::Uload16x4
+            | Opcode::Uload32x2
+            | Opcode::Sload8x8
+            | Opcode::Sload16x4
+            | Opcode::Sload32x2 => 8,
+            _ => dfg.ctrl_typevar(inst).bytes() as i64,
+        };
+
+        let offset: i64 = offset.into();
+
+        // A negative index directly into a `heap_addr` is always illegal.
+        if offset < 0 {
+            return errors.fatal((
+                inst,
+                format!("offset cannot be negative when referencing a heap_addr, got {offset}"),
+            ));
+        }
+
+        // Check if the access is within the bounds provided by  heap_addr
+        let oob_bytes = -(valid_size - (offset + access_size));
```

Comment:
This is algebraically equivalent to what you wrote, but makes more sense to me:

```suggestion
        let needed_size = offset + access_size;
        let oob_bytes = needed_size - valid_size;
```